### PR TITLE
Fixes #34823 - Upsert features when registering them

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Review by ares — COMMENTED** [review]
It all looks reasonable, I'd like to better understand it though. Please ping me when you answer, I'm not watching gh notifications at the moment.

File: app/models/remote_execution_feature.rb [hunk 1]
```diff
@@ -43,8 +43,12 @@ def self.register(label, name, options = {})
     feature = self.find_by(label: label)
     builder = options[:notification_builder] ? options[:notification_builder].to_s : nil
 
+    if options[:provided_inputs]
+      provided_inputs = Array(options[:provided_inputs]).join(',')
+    end
+
     attributes = { :name => name,
-                   :provided_input_names => options[:provided_inputs],
+                   :provided_inputs => provided_inputs,
```
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/models/remote_execution_feature.rb:51]: It is. During upsert the provided attributes do not go through the regular model and association machinery so the format needs to be changed in bit in order for it to go through.

File: app/models/remote_execution_feature.rb [hunk 2]
```diff
@@ -58,13 +62,12 @@ def self.register(label, name, options = {})
...
-        feature = self.create!({ :label => label }.merge(attributes))
-      else
-        feature.attributes = attributes
-        feature.save if feature.changed?
-      end
-      return feature
+      # The only validation we currently have is uniqueness validation, which
+      # the upsert will enforce
+      # rubocop:disable Rails/SkipsModelValidations
+      result = self.upsert({ label: label }.merge(attributes), unique_by: :label)
```
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/models/remote_execution_feature.rb:68]: > what happens if this fails?

How would this fail?

> and is the key that this atomically updates or creates the record?

Correct

> was the cause race condition with the if?

I don't really remember, it's been a while, but most likely yes.

File: db/migrate/20220426145007_add_unique_feature_label_index.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class AddUniqueFeatureLabelIndex < ActiveRecord::Migration[6.0]
+  def change
+    remove_index :remote_execution_features, :label
+    add_index :remote_execution_features, :label, unique: true
```
[the remark this answers is not in the document]
Comment (adamruzicka) [reply db/migrate/20220426145007_add_unique_feature_label_index.rb:12]: Changed it to prune the duplicate records so there's only one left. Once the initializers kick in, the last remaining record (for each label) should get updated

**Review by adamruzicka — COMMENTED** [review]
Hopefully answered the questions and rebased for a good measure

File: app/models/remote_execution_feature.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ares) [reply app/models/remote_execution_feature.rb:68]: hm, upsert does not call validations and callbacks, so I guess if it fails, something is wrong with the DB

**Review by ares — APPROVED** [review]
Thanks @adamruzicka, merging

**Pull request merged**